File: ts/src/registry.ts
import { listSpecDir } from "./spec";
import { loadHostFile, type Host } from "./host";

interface Registry {
  byName: Map<string, Host>;
  aliasIndex: Map<string, string>;
}

let cached: Registry | undefined;

function loadRegistry(): Registry {
  const byName = new Map<string, Host>();
  const aliasIndex = new Map<string, string>();

  for (const entry of listSpecDir("hosts")) {
    const host = loadHostFile(`hosts/${entry}/host.yaml`);
    byName.set(host.name, host);
    for (const alias of host.aliases) {
      aliasIndex.set(alias, host.name);
    }
  }

  return { byName, aliasIndex };
}

function registry(): Registry {
  if (cached === undefined) cached = loadRegistry();
  return cached;
}

/** Every host, sorted by canonical name. */
export function hosts(): Host[] {
  return [...registry().byName.values()].sort((a, b) => a.name.localeCompare(b.name));
}

/** Looks up a host by canonical name only -- never an alias. */
export function get(name: string): Host | undefined {
  return registry().byName.get(name);
}

/**
 * Accepts a canonical name or a published alias. The only alias-aware
 * function in this module.
 */
export function resolve(nameOrAlias: string): Host | undefined {
  const direct = registry().byName.get(nameOrAlias);
  if (direct) return direct;
  const canonical = registry().aliasIndex.get(nameOrAlias);
  return canonical === undefined ? undefined : registry().byName.get(canonical);
}

/** Hosts with a hook surface (hooks: true). */
export function hookedHosts(): Host[] {
  return hosts().filter((h) => h.hooks);
}

File: go/registry.go
package axon

import (
	"io/fs"
	"sort"
	"sync"
)

var (
	hostsOnce   sync.Once
	hostsByName map[string]Host
	aliasIndex  map[string]string
	hostsErr    error
)

func loadHosts() {
	hostsByName = map[string]Host{}
	aliasIndex = map[string]string{}
	entries, err := fs.ReadDir(Spec(), "hosts")
	if err != nil {
		hostsErr = err
		return
	}
	for _, e := range entries {
		if !e.IsDir() {
			continue
		}
		var h Host
		if err := loadYAML(Spec(), "hosts/"+e.Name()+"/host.yaml", "host.schema.json", &h); err != nil {
			hostsErr = err
			return
		}
		hostsByName[h.Name] = h
		for _, a := range h.Aliases {
			aliasIndex[a] = h.Name
		}
	}
}

func ensureHosts() {
	hostsOnce.Do(loadHosts)
	if hostsErr != nil {
		panic(hostsErr)
	}
}

// Hosts returns every host sorted by canonical name.
func Hosts() []Host {
	ensureHosts()
	out := make([]Host, 0, len(hostsByName))
	for _, h := range hostsByName {
		out = append(out, h)
	}
	sort.Slice(out, func(i, j int) bool { return out[i].Name < out[j].Name })
	return out
}

// Get looks up a host by canonical name only.
func Get(name string) (Host, bool) {
	ensureHosts()
	h, ok := hostsByName[name]
	return h, ok
}

// Resolve accepts a canonical name or a published alias. It is the only
// alias-aware function in the module.
func Resolve(nameOrAlias string) (Host, bool) {
	ensureHosts()
	if h, ok := hostsByName[nameOrAlias]; ok {
		return h, true
	}
	if canon, ok := aliasIndex[nameOrAlias]; ok {
		return hostsByName[canon], true
	}
	return Host{}, false
}

// HookedHosts returns hosts with a hook surface (hooks: true).
func HookedHosts() []Host {
	var out []Host
	for _, h := range Hosts() {
		if h.Hooks {
			out = append(out, h)
		}
	}
	return out
}
